### PR TITLE
refactor: extract testable prompt helpers from the two judges

### PR DESCRIPTION
**Pull request by prasanth-nair-kv**
## What & why

`judgeResponse` (~115 lines) and `judgeToolResponse` each bundled prompt
assembly with the LLM call, so their subtle logic was only reachable through a
live judge run. This extracts two **behavior-preserving, unit-testable** helpers
(from the PR4 max-review, findings #6/#7):

- **`pairTurnsForJudge`** (`evaluators/judge.ts`) — the role-alternation pairing
  window: filter to user/assistant, step-by-2, warn-and-fallback to a single
  synthetic turn.
- **`buildMcpJudgePrompt` + `McpJudgePromptInput`** (`run/judge.ts`) — the MCP
  judge user-prompt assembly (description-scan branch, prior-turns rendering,
  evidence-rule ordering). `judgeToolResponse` is now just build → call → parse.

## Behavior-preserving

Pure mechanical moves — the prompt arrays were copied verbatim. Guards:
- an **exact-string snapshot** pins the MCP prompt byte-for-byte;
- focused unit tests for the pairing helper (clean pairing, fallback on
  undefined/empty/single, trailing-unpaired-turn dropped, `_opfor_*` arg stripping,
  each prompt branch);
- the `runAll.smoke` + `orchestrator.equivalence` integration guards stay green.

`judgeToolResponse`'s args type moved from an inline object to
`McpJudgePromptInput & { model }` — structurally identical, and typecheck confirms
no caller breaks.

## Verification

typecheck 0 errors · lint 0 issues · full suite 82 tests, 0 fail (3 skips) · build
(incl. esbuild browser bundle) passes.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Improved judge prompt generation for tool responses and conversation history, making turn pairing and prompt assembly more consistent.
  * Added clearer handling for tool errors, scan/description mode, and prior turns in judge prompts.

* **Bug Fixes**
  * Falls back safely when conversation history cannot form valid turn pairs.
  * Preserves the strongest judge instructions and removes internal-only fields from the rendered prompt.

* **Tests**
  * Added coverage for prompt rendering and turn pairing across common and edge-case scenarios.

### CODE REVIEW

File: core/src/evaluators/judge.ts [hunk 1]
```diff
@@ -48,6 +48,41 @@ export interface AttackContext {
...
+  if (turns.length === 0) {
+    if (conversationHistory && conversationHistory.length >= 2) {
+      log.warn(
+        `judge: conversationHistory had ${conversationHistory.length} entries but yielded zero user/assistant pairs — falling back to single-turn block. Check role alternation in the caller.`
+      );
+    }
+    turns.push(fallback);
+  }
+  return turns;
+}
```
[the remark this answers is not in the document]
Comment (jithin23-kv) [reply core/src/evaluators/judge.ts:105]: Fixed. Replaced the fixed step-by-2 loop with a greedy pass that resyncs by one entry on a role mismatch, so a valid user/assistant pair after a misaligned entry is still captured instead of dropped. Added a `resynced` flag to detect genuine misalignment (distinct from a benign trailing odd turn) and warn on it. Added a regression test covering `user/assistant/assistant/user/assistant`.

**Comment by jithin23-kv** [comment]
Addressed the review findings:

- **Turn pairing** — `pairTurnsForJudge` now pairs greedily and resyncs by one entry on a role mismatch, so a valid user/assistant pair after a misaligned entry is captured instead of silently dropped. Warns on genuine misalignment.
- **MCP scan input** — the description-scan response is now validated with a Zod schema instead of a `JSON.parse` cast (per the "Zod for all external input" convention), falling back to the raw response on invalid/non-JSON input.
- **Cleanup** — extracted the repeated `{ user, assistant }` shape into a `JudgeTurn` alias.
- **Test** — added a regression test covering resync recovery on a misaligned transcript.

File: core/src/evaluators/judge.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (jithin23-kv) [reply core/src/evaluators/judge.ts:105]: pushed changes now

**Review by jithin23-kv — APPROVED** [review]
